### ISSUE TITLE
The implementation ignores URLs on different subdomains due to urlIsLegal implementation.

### ISSUE THREAD

**Issue by suhdev** [crawler-commons/crawler-commons#81]
The current implementation assumes that a URL is legal only if the sitemap URL is a substring of the URL. This doesn't hold for some websites such as nytimes.com in which the sitemaps are actually on a different subdomain and the urls within those sitemaps points to pages on the main domain. 

```
Snippet from nytimes.com robots.txt: 

Sitemap: http://spiderbites.nytimes.com/sitemaps/www.nytimes.com/sitemap.xml.gz
Sitemap: http://www.nytimes.com/sitemaps/sitemap_news/sitemap.xml.gz
Sitemap: http://spiderbites.nytimes.com/sitemaps/sitemap_video/sitemap.xml.gz

```

Perhaps using the domain name, not the whole sitemap URL can fix this, Google Guava `InternetDomainName` which could help in extracting domain names, or maybe using simple Regex!.

**Comment by kkrugler** [comment]
From http://www.sitemaps.org/protocol.html#location:

> The location of a Sitemap file determines the set of URLs that can be included in that Sitemap. A Sitemap file located at http://example.com/catalog/sitemap.xml can include any URLs starting with http://example.com/catalog/ but can not include URLs starting with http://example.com/images/.

and then:

>  Note that this means that all URLs listed in the Sitemap must use the same protocol (http, in this example) and reside on the same host as the Sitemap. For instance, if the Sitemap is located at http://www.example.com/sitemap.xml, it can't include URLs from http://subdomain.example.com.

So it seems like what you're asking for is support for invalid/broken sitemaps, yes?

Dealing with invalid data (robots.txt, etc) is definitely something we do a lot of, but I think this would want to be handled via a "lenient" vs. "strict" setting.

**Comment by Chaiavi** [comment]
Actually, nytimes are ok, thier sitemap isn't broken.

You can have in your sitemap file, "cross-domain" urls ONLY IF you
authorize those domains in your robots.txt file.


This specific case is mentioned in the same link under the Sitemaps & Cross
Submits subtitle.




This is a feature we don't support yet (although we should consider it
IMHO), it involves parsing the robots.txt file when parsing a sitemap.

**Comment by kkrugler** [comment]
Hi Avi - thanks for the ref...cross submits were added after the original sitemap spec, so I didn't know about them.

I'm still a bit confused about the nytimes.com case. Based on this:

> To submit Sitemaps for multiple hosts from a single host, you need to "prove" ownership of the host(s) for which URLs are being submitted in a Sitemap

The "Sitemaps & Cross Submits" is for when you want to have a combo sitemap (on say the www.nytimes.com domain) that has URLs for spiderbites.nytimes.com. So then you'd modify the spiderbites.nytimes.com/robots.txt file to have a sitemap URL that points to www.nytimes.com, which "proves" that you have control over the spiderbites.nytimes.com domain, and thus it's OK to have those URLs in a sitemap on another domain.

But here the case is different - it's not about having a combo sitemap file, it's about having distributed sitemaps on subdomains that can be found from a parent robots.txt.

Is there another section of the spec that covers this case?

Also note that parsing the robots.txt file to validate the sitemap isn't sufficient for the combo sitemap case - you need to get the robots.txt files from all of the "other" domains, and verify that they have references to sitemaps on the domain with the sitemap that's being parsed.

**Comment by Chaiavi** [comment]
Inline

On Sun, Jun 7, 2015 at 6:32 PM, Ken Krugler <notifications@github.com>
wrote:

> Hi Avi - thanks for the ref...cross submits were added after the original
> sitemap spec, so I didn't know about them.
>
> I'm still a bit confused about the nytimes.com case. Based on this:
>
> To submit Sitemaps for multiple hosts from a single host, you need to
> "prove" ownership of the host(s) for which URLs are being submitted in a
> Sitemap
>
> The "Sitemaps & Cross Submits" is for when you want to have a combo
> sitemap (on say the www.nytimes.com domain) that has URLs for
> spiderbites.nytimes.com. So then you'd modify the
> spiderbites.nytimes.com/robots.txt file to have a sitemap URL that points
> to www.nytimes.com, which "proves" that you have control over the
> spiderbites.nytimes.com domain, and thus it's OK to have those URLs in a
> sitemap on another domain.
>
[Avi] Exactly


> But here the case is different - it's not about having a combo sitemap
> file, it's about having distributed sitemaps on subdomains that can be
> found from a parent robots.txt.
>
> Is there another section of the spec that covers this case?
>
[Avi] It is the same case and the same spec, just the perspective is the
opposite one.
In nytimes case, they wanted to have sitemaps on the subdomains pointing to
the main site, so they added the Sitemap ref on the main domain to point to
the subdomains for authentication purposes, now the subdomains can point to
urls on the main site.

Their scenario is legal it just isn't the regular one, one would expect the
main site to have sitemaps pointing to urls on its subdomains but they
decided to do the opposite thing and have sitemaps on subdomains pointing
to the main site.
I assume they have their SEO calculations there and their machine generated
sitemaps made a mess of things.


> Also note that parsing the robots.txt file to validate the sitemap isn't
> sufficient for the combo sitemap case - you need to get the robots.txt
> files from all of the "other" domains, and verify that they have references
> to sitemaps on the domain with the sitemap that's being parsed.
>
[Avi] You are right.
And there is another limitation - every sitemap can contain URLs to a
SINGLE domain only! (if it is authenticated then it can contain URLs to the
other domain but only to that domain):
When a particular host's robots.txt, say http://www.host1.com/robots.txt,
points to a Sitemap or a Sitemap index on another host; it is expected that
for each of the target Sitemaps, such as
http://www.sitemaphost.com/sitemap-host1.xml, all the URLs belong to the
host pointing to it. This is because, as noted earlier, a Sitemap is
expected to have URLs from a single host only.

**Comment by suhdev** [comment]
Insightful.

@kkrugler  I knew about the "strict" flag, but the URLs seemed legal for me, the sitemaps are included in the `nytimes.com/robots.txt`, and it seemed like they the whole (spiderbytes.nytimes.com) is just an index with urls pointing to every article on the nytimes.com. I assumed they're trying to take the load of serving all those sitemaps. 

I came across this problem while building a crawler, and I was actually running unit tests against the nytimes sitemap (the sitemap is big enough), for memory usage tests. 

Thank you all, and I've forked the repository, and I'll look into following @Chaiavi solution's of adding the robots.txt check to the implementation. Maybe having the `robots.txt` parser to generate a set of whitelisted domains that can then be injected into the sitemap parser. However, for now, I'll be using the 'lenient' mode.
